core/drivers/crypto/se050/core/rsa.c

```diff
@@ -0,0 +1,705 @@
+	return TEE_ERROR_OUT_OF_MEMORY;
+}
+
+static void do_free_publickey(struct rsa_public_key *s)
+{
+	if (!s)
+		return;
+	crypto_bignum_free(s->n);
+	crypto_bignum_free(s->e);
+}
+
+static void do_free_keypair(struct rsa_keypair *s)
+{
+	sss_status_t st = kStatus_SSS_Fail;
+	sss_se05x_object_t k_object = { 0 };
+	uint32_t key_id;
+
+	key_id = se050_rsa_keypair_from_nvm(s);
+	if (key_id) {
+		st = sss_se05x_key_object_get_handle(&k_object, key_id);
+		if (st != kStatus_SSS_Success) {
+			EMSG("RSA ");
+			return;
+		}
+		sss_se05x_key_store_erase_key(se050_kstore, &k_object);
+	}
+
+	crypto_bignum_free(s->e);
+	crypto_bignum_free(s->d);
+	crypto_bignum_free(s->n);
+	crypto_bignum_free(s->p);
+	crypto_bignum_free(s->q);
+	crypto_bignum_free(s->qp);
+	crypto_bignum_free(s->dp);
+}
+
+static TEE_Result do_gen_keypair(struct rsa_keypair *key, size_t kb)
+{
+	sss_status_t st = kStatus_SSS_Fail;
+	sss_se05x_object_t k_object = { 0 };
+	TEE_Result ret = TEE_SUCCESS;
+	uint32_t oid = 0;
+	uint64_t kid = 0;
+	uint8_t k[2048] = { 0 };
+	uint8_t *n = NULL;
+	uint8_t *e = NULL;
+	size_t n_len = 0, e_len = 0, k_len = sizeof(k);
+
+	st = sss_se05x_key_object_init(&k_object, se050_kstore);
+	if (st != kStatus_SSS_Success)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	st = se050_get_oid(kKeyObject_Mode_Persistent, &oid);
+	if (st != kStatus_SSS_Success)
+		return TEE_ERROR_GENERIC;
+
+	st = sss_se05x_key_object_allocate_handle(&k_object, oid,
+						  kSSS_KeyPart_Pair,
+						  kSSS_CipherType_RSA, 0,
+						  kKeyObject_Mode_Persistent);
+	if (st != kStatus_SSS_Success)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	st = sss_se05x_key_store_generate_key(se050_kstore, &k_object, kb,
+					      &se050_asym_policy);
+	if (st != kStatus_SSS_Success) {
+		sss_se05x_key_store_erase_key(se050_kstore, &k_object);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	st = sss_se05x_key_store_get_key(se050_kstore, &k_object, k, &k_len,
+					 &kb);
+	if (st != kStatus_SSS_Success) {
+		ret = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
+	}
+
+	st = sss_util_asn1_rsa_parse_public(k, k_len, &n, &n_len, &e, &e_len);
+	if (st != kStatus_SSS_Success) {
+		ret = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
+	}
+
+	crypto_bignum_bin2bn(n, n_len, key->n);
+	crypto_bignum_bin2bn(e, e_len, key->e);
+
+	/* the se050 does not provide the private key */
+	kid = se050_generate_private_key(oid);
+	crypto_bignum_bin2bn((uint8_t *)&kid, sizeof(kid), (key->d));
+
+	/* fill the rest with some data */
+	crypto_bignum_bin2bn((uint8_t *)&oid, sizeof(oid), key->p);
+	crypto_bignum_bin2bn((uint8_t *)&oid, sizeof(oid), key->q);
+	crypto_bignum_bin2bn((uint8_t *)&oid, sizeof(oid), key->qp);
+	crypto_bignum_bin2bn((uint8_t *)&oid, sizeof(oid), key->dp);
+	crypto_bignum_bin2bn((uint8_t *)&oid, sizeof(oid), key->dq);
+
+	free(n);
+	free(e);
+
+	ret = TEE_SUCCESS;
+exit:
+	if (ret != TEE_SUCCESS) {
+		IMSG("RSA key generation failed");
+		sss_se05x_key_store_erase_key(se050_kstore, &k_object);
+	}
+
+	return TEE_SUCCESS;
```

Comment:
`return ret`

Reply:
absolutely,. thanks for catching that one!